packages/flutter_test/test/widget_tester_test.dart

```diff
@@ -699,6 +701,18 @@ void main() {
       ticker.stop();
   });
 
+  group('testWidgets matrix works', () {
+    int numberOfVariationsRun = 0;
+
+    testWidgets('matrix tests run all variations', (WidgetTester tester) async {
+      if (debugDefaultTargetPlatformOverride == null) {
+        expect(numberOfVariationsRun, equals(TargetPlatform.values.length));
+      } else {
+        numberOfVariationsRun += 1;
+      }
+    }, matrix: <TestDimension<dynamic>>[ TargetPlatformDimension(TargetPlatform.values), DefaultDimension() ]);
```

Comment:
"matrix" and "dimension" really imply that this is going to do the full combination. It seems weird that they are treated as one dimension and that it's a vector rather than an actual matrix.

Reply:
I don't think that's accurate:

| | android | iOS | 800x600 | 640x480 |
|-|-|-|-|-|
| target platform | | | | |
| screen size | | | | |

It's just one dimension, with multiple units.

I really don't know how useful that really is. If you're going to accept multiple values here, I don't see any reason you _wouldn't_ want to do the full combination matrix.

Reply:
Because it doesn't give you as much control over which tests are run.

For instance, if I wanted to (for the same test) also run variants where `debugInstrumentationEnabled` was `true` and `false`, along with `TargetPlatform.android`, `TargetPlatform.iOS` and `TargetPlatform.macOS`, but I knew that there wasn't any interaction between the two variables (e.g. maybe I'm just benchmarking?), then it's kind of pointless to run each instrumentation test three times with the different platforms (a total of six tests), when I actually only want to run five tests (each of the three platforms, and then one each of the instrumentation tests with the default value of target platform).  If I *did* want to run the full matrix, then I could just make a `TestVariant` that enumerates all six states, and they would all be run.

Reply:
But when would that be the case?

Also, that would run 5 cases but actually it would run 4 cases, then duplicate one of the four. If you're doing 5 cases and wasting time on one of them, why not do 6 cases?

It just seems like all the use cases I would actually imagine, where one test is interesting for two different axes, would almost by definition be cases where the combination of the axes is even more interesting.

If we only support one axis, then I think we should just take one test dimension (and definitely not call it "matrix"... you could call it "dimension" or "domain"), and just run it on that. With only one axis, it seems unlikely that you'd ever actually care about combining unrelated types.

Reply:
The word I'm using is "variant", and each "variant" can take on multiple "values".

Is that OK terminology?

Reply:
Ian and I spoke offline and decided that we could just implement this as a single object instead of as a list.